### ISSUE TITLE
Unwrap_used false positive with Infallible return value

### ISSUE THREAD

**Issue by sww1235** [rust-lang/rust-clippy#16709]
### Summary

Unwrapping a function that returns `core::convert::Infallible` or `std::convert::Infallible` results in a false positive `clippy::unwrap_used` lint.

May be related to #11245

### Lint Name

unwrap_used

### Reproducer

First I tried this, in a function that returns `Result<Svg, Error>` where `Error` is a custom struct based Error type.

```rust
let output = Svg::from_str(&svg_string)?;
```
Which resulted in a error:

```
error[E0277]: `?` couldn't convert the error to `error::Error`
   --> crates/cdm_core/src/datatypes/project_types/mounting_rail.rs:106:48
    |
106 |         let output = Svg::from_str(&svg_string)?;
    |                      --------------------------^ the trait `std::convert::From<std::convert::Infallible>` is not implemented for `error::Error`
    |                      |
    |                      this can't be annotated with `?` because it has type `Result<_, std::convert::Infallible>`
    |
note: `error::Error` needs to implement `From<std::convert::Infallible>`
   --> crates/cdm_core/src/error.rs:11:1
```

Fair enough: I tried this to work around the issue:

```rust
let output = Svg::from_str(&svg_string).unwrap();
```
Which triggered the following lints:
```
// at the call site
clippy::unwrap_used

// on the containing function
clippy::missing_panics_doc
```

then I tried adding an `#[expect]`:

```rust
#[expect(clippy::unwrap_used, reason = "this function in infalliable")]
let output = Svg::from_str(&svg_string).unwrap();
```
which triggered the following lints:
```
// at the call site
clippy::unwrap_used

// on the containing function
clippy::missing_panics_doc

// on the expect attribute
this lint expectation is unfulfilled
```

I expected to see this happen:

No warning or error while unwrapping the return value of a function that returns `core::convert::Infallible` or `std::convert::Infallible` per the fixes in #11245

------
`FromStr` definition for reference. `Infallible` is `core::convert::Infallible` here, but I have also tried `std::convert::Infallible` with no changes.

```rust
impl FromStr for Svg {
    type Err = Infallible;

    //TODO: validation
    /// Creates a `Svg` from a string representation of an Svg. No validation is performed
    /// currently.
    #[inline]
    fn from_str(text: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            svg_data: text.to_owned(),
            filepath: None,
        })
    }
}
```

### Version

```text
rustc 1.94.0 (4a4ef493e 2026-03-02)
binary: rustc
commit-hash: 4a4ef493e3a1488c6e321570238084b38948f6db
commit-date: 2026-03-02
host: x86_64-unknown-linux-gnu
release: 1.94.0
LLVM version: 21.1.8
```

**Comment by sww1235** [comment]
Source code links if it helps:

https://github.com/sww1235/connection-diagram-manager/blob/gui-dev/crates/cdm_core/src/datatypes/project_types/mounting_rail.rs

https://github.com/sww1235/connection-diagram-manager/blob/gui-dev/crates/cdm_core/src/datatypes/svg.rs

**Comment by sww1235** [comment]
May also be related to #12998 or #9361. In this case, I don't think it quite applies to #12998 as the `#[expect]` is not actually suppressing the lint but they popped up while googling so I figured I would add them here as references.

**Comment by sww1235** [comment]
just had a similar issue occur with `clippy::wildcard_enum_match_arm`. Let me know if you want me to file a separate issue ticket for that.

**Comment by zihan0822** [comment]
in my local reproduction, `unwrap_in_result` is fired instead of `unwrap_used`. you can try to suppress it with `#[expect(clippy::unwrap_in_result)]`. infallible result has actually been handled properly in `unwrap_used`, but not yet in `unwrap_in_result`. i just submitted a PR to fix this. we properly need to handle this in panic doc lint as well.

**Comment by sww1235** [comment]
Assuming you mean `Infallible` wasn't handled properly in `unwrap_in_result`?

Changing from `unwrap_used` to `unwrap_in_result` in my `expect` did indeed suppress both `lint_expectation_is_unfulfilled` and `unwrap_used` lints being triggered. Not sure why the original message still complains about `unwrap_used`. 

I am still getting the issue with the panic doc lint, but that seems expected at this point, given your last sentence.

Thank you!
